src/js/sonifier.js

```diff
@@ -0,0 +1,402 @@
+    // takes by summing note duration, but we can't receive event notification
+    // when a sonification completes
+    // - we can fire an event when a voice label read completes, but can't know
+    // in advance how long it will take to read the label
+    floe.chartAuthoring.sonifier.processSonificationQueue = function(delay, noGap, that) {
+        that.applier.change("isPlaying", true);
+        var gapDuration = that.options.playbackOptions.gapDuration;
+
+        var sonificationQueue = that.model.sonificationQueue;
+
+        var synth = that.model.synth;
+
+        var currentData = sonificationQueue[0];
+
+        var textToSpeech = that.textToSpeech;
+
+        // Schedule the next voice label, accounting for both the variable-length
+        // delay (the time to play the preceding sonification) and the fixed-length
+        // gap
+
+        // We shouldn't use the gap if this is the first call for a dataset
+        var pause = noGap? delay : delay+gapDuration;
+        if(fluid.textToSpeech.isSupported()) {
+            synth.scheduler.once(pause, function() {
+                that.applier.change("currentlyPlayingData", currentData);
+                textToSpeech.queueSpeech(currentData.label);
+            });
+        }
+    };
+
+    // Recursion function called from floe.chartAuthoring.sonifier.processSonificationQueue
+
+    floe.chartAuthoring.sonifier.playDataAndQueueNext = function(that) {
+
+        var synth = that.model.synth;
+        var sonificationQueue = that.model.sonificationQueue;
+        var data = sonificationQueue.shift();
+
+        // This prevents an error being thrown if a stop has been called but
+        // this function is queued or in the midst of execution
+        var noteDuration = (data !== undefined) ? floe.chartAuthoring.sonifier.getTotalDuration(data.notes.durations) : 0;
+
+        if(sonificationQueue.length > 0) {
+            floe.chartAuthoring.sonifier.processSonificationQueue(noteDuration, false, that);
+        } else {
+            // Schedule to stop the sonification after the last sonification is
+            // played
+            synth.scheduler.once(noteDuration, function() {
+                that.stopSonification();
+            });
+        }
+        if(data !== undefined) {
+            synth.midiNoteSynth.applier.change("inputs.noteSequencer", data.notes);
+            synth.pianoEnvelopeSynth.applier.change("inputs.envelopeSequencer", data.envelope);
+        }
+    };
+
+    // Given an array containing indiividual durations, accumulate them and
+    // return the total duration
+    floe.chartAuthoring.sonifier.getTotalDuration = function (durationsArray) {
+        var sum = function(duration, runningTotal) {
+            return duration+runningTotal;
+        };
+        return fluid.accumulate(durationsArray, sum, 0);
+    };
+
+    floe.chartAuthoring.sonifier.stopSonification = function(that) {
+
+        if(that.model.isPlaying) {
+            try {
```

Comment:
Why is this wrapped in a <code>try/catch/finally</code>? Which of these functions will potentially throw, and under what circumstances?